engines/scumm/he/net/net_main.cpp

```diff
@@ -0,0 +1,1447 @@
+
+	if (_isHost) {
+		if (getTotalPlayers() > 4) {
+			// We are full.
+			return 0;
+		}
+		_userIdToName[++_userIdCounter] = longName;
+		_numUsers++;
+		if (_sessionId && _sessionServerPeer > -1) {
+			// Update player count to session server {
+			Common::String updatePlayers = Common::String::format(
+				"{\"cmd\":\"update_players\",\"game\":\"%s\",\"version\":\"%s\",\"players\":%d}",
+				_gameName.c_str(), _gameVersion.c_str(), getTotalPlayers());
+			_sessionHost->send(updatePlayers.c_str(), _sessionServerPeer);
+		}
+		return 1;
+	}
+
+	// Client:
+	if (_myUserId != -1)
+		return 1;
+
+	Common::String addUser = Common::String::format(
+		"{\"cmd\":\"add_user\",\"name\":\"%s\"}", longName);
+
+	_sessionHost->send(addUser.c_str(), 0, 0, true);
+
+	uint tickCount = 0;
+	while(_myUserId == -1) {
+		remoteReceiveData(12);
+		// Wait for five seconds for our user id before giving up
+		tickCount += 5;
+		g_system->delayMillis(5);
+		if (tickCount >= 5000)
+			break;
+	}
+	return (_myUserId > -1) ? 1 : 0;
+}
+
+int Net::removeUser() {
+	debug(1, "Net::removeUser()"); // PN_RemoveUser
+
+	if (_myUserId != -1)
+		destroyPlayer(_myUserId);
+
+	return 1;
+}
+
+int Net::whoSentThis() {
+	debug(1, "Net::whoSentThis(): return %d", _fromUserId); // PN_WhoSentThis
+	return _fromUserId;
+}
+
+int Net::whoAmI() {
+	debug(1, "Net::whoAmI(): return %d", _myUserId); // PN_WhoAmI
+	return _myUserId;
+}
+
+int Net::createSession(char *name) {
+	debug(1, "Net::createSession(\"%s\")", name); // PN_CreateSession
+
+	if (!_enet) {
+		return 0;
+	};
+
+	_sessionId = -1;
+	_sessionName = name;
+	// Normally we would do only one peer (0) or three peers (2) but we are reserving one
+	// for our connection to the session server.
+	_sessionHost = _enet->createHost("0.0.0.0", 0, _maxPlayers);
+
+	if (!_sessionHost) {
+		return 0;
+	}
+
+	_isHost = true;
+
+	bool enableSessionServer = true;
+	bool enableLanBroadcast = true;
+	if (ConfMan.hasKey("enable_session_server"))
+		enableSessionServer = ConfMan.getBool("enable_session_server");
+	if (ConfMan.hasKey("enable_lan_broadcast"))
+		enableLanBroadcast = ConfMan.getBool("enable_lan_broadcast");
+
+	if (enableSessionServer) {
+		if (!_forcedAddress && ConfMan.hasKey("session_server")) {
+			_sessionServerAddress = getAddressFromString(ConfMan.get("session_server"));
+			// Set port to default if not defined.
+			if (!_sessionServerAddress.port)
+				_sessionServerAddress.port = 9120;
+		}
+		if (_sessionHost->connectPeer(_sessionServerAddress.host, _sessionServerAddress.port)) {
+			// FIXME: Get the IP address of the session server when a domain address is used.
+
+			// _sessionServerPeer = _sessionHost->getPeerIndexFromHost(_sessionServerAddress.host, _sessionServerAddress.port);
+			_sessionServerPeer = 0;
+			// Create session to the session server.
+			Common::String req = Common::String::format(
+				"{\"cmd\":\"host_session\",\"game\":\"%s\",\"version\":\"%s\",\"name\":\"%s\",\"maxplayers\":%d}",
```

Comment:
I wonder, is it absolutely necessary to use a JSON object here as you mention in your TODO? I think, the code will go like 5x with JSON.